File: scripts/q1_plots.jl
#!/usr/bin/env julia
# Q1 plotting: reads the four Q1 CSVs and writes PNGs into figures/.
#
#   q1-angle-schedules.csv      → figures/q1-angle-schedules.png   (E1)
#   q1-intermediate-depth.csv   → figures/q1-intermediate-depth.png (E2)
#   q1-adiabatic-init.csv       → figures/q1-adiabatic-init.png    (E3)
#   q1-angle-curvature.csv      → figures/q1-angle-curvature.png   (E4)
#
# Usage: julia --project=. scripts/q1_plots.jl
#
# Requires Plots.jl. If not available the script aborts with a clear message
# rather than mutating Project.toml.

using Printf, DelimitedFiles

const RESULTS_DIR = joinpath(@__DIR__, "..", "results")
const FIGURES_DIR = joinpath(@__DIR__, "..", "figures")
mkpath(FIGURES_DIR)

try
    @eval using Plots
catch e
    @error "Plots.jl is required to render Q1 figures. Install it once with " *
           "`julia --project=. -e 'using Pkg; Pkg.add(\"Plots\")'`."
    rethrow(e)
end

# ── Tiny CSV reader (skip lines starting with '#') ──────────────────────
function load_csv(path)
    lines = filter(l -> !startswith(l, '#') && !isempty(l), readlines(path))
    header = split(popfirst!(lines), ',')
    rows = [split(l, ',') for l in lines]
    return header, rows
end

function getcol(header, rows, name)
    idx = findfirst(==(name), header)
    idx === nothing && error("column $name not in $(header)")
    return [r[idx] for r in rows]
end

asfloat(xs) = parse.(Float64, xs)
asint(xs)   = parse.(Int,     xs)

# ── E1: angle schedules ────────────────────────────────────────────────

function plot_angle_schedules()
    path = joinpath(RESULTS_DIR, "q1-angle-schedules.csv")
    isfile(path) || (println("skip E1 (no $path)"); return)
    header, rows = load_csv(path)

    Ds = unique(asint(getcol(header, rows, "D")))
    γ_plt = plot(title="γ schedule: optimal vs linear adiabatic",
                 xlabel="step j / p", ylabel="γ", legend=:outertopright)
    β_plt = plot(title="β schedule: optimal vs linear adiabatic",
                 xlabel="step j / p", ylabel="β", legend=:outertopright)

    for D in sort(Ds)
        sel = [i for (i, r) in enumerate(rows) if parse(Int, r[2]) == D]
        isempty(sel) && continue
        sub = [rows[i] for i in sel]
        p   = parse(Int, sub[1][3])
        j   = asint(getcol(header, sub, "j"))
        x   = j ./ p
        γo  = asfloat(getcol(header, sub, "gamma_opt"))
        βo  = asfloat(getcol(header, sub, "beta_opt"))
        γa  = asfloat(getcol(header, sub, "gamma_adi"))
        βa  = asfloat(getcol(header, sub, "beta_adi"))
        plot!(γ_plt, x, γo, label="D=$D opt", lw=2, marker=:circle)
        plot!(γ_plt, x, γa, label="D=$D adi", ls=:dash)
        plot!(β_plt, x, βo, label="D=$D opt", lw=2, marker=:circle)
        plot!(β_plt, x, βa, label="D=$D adi", ls=:dash)
    end

    fig = plot(γ_plt, β_plt, layout=(1, 2), size=(1400, 500))
    out = joinpath(FIGURES_DIR, "q1-angle-schedules.png")
    savefig(fig, out)
    println("wrote $out")
end

# ── E2: intermediate-depth performance ─────────────────────────────────

function plot_intermediate_depth()
    path = joinpath(RESULTS_DIR, "q1-intermediate-depth.csv")
    isfile(path) || (println("skip E2 (no $path)"); return)
    header, rows = load_csv(path)

    Ds = sort(unique(asint(getcol(header, rows, "D"))))
    plt = plot(title="QAOA intermediate-depth performance",
               xlabel="depth t", ylabel="c̃",
               legend=:bottomright, ylim=(0.45, 0.92))

    for D in Ds
        sub = filter(r -> parse(Int, r[1]) == D, rows)
        t   = asint(getcol(header, sub, "t"))
        ct  = asfloat(getcol(header, sub, "ctilde_truncated"))
        co  = asfloat(getcol(header, sub, "ctilde_optimal_at_t"))
        ca  = asfloat(getcol(header, sub, "ctilde_adiabatic_at_t"))
        plot!(plt, t, co, label="D=$D optimal@t", lw=2,  marker=:circle)
        plot!(plt, t, ct, label="D=$D truncated", lw=1,  ls=:dash, marker=:utriangle)
        plot!(plt, t, ca, label="D=$D adiabatic", lw=1,  ls=:dot,  marker=:diamond)
    end
    hline!(plt, [0.5], color=:gray, ls=:dot, label="random (c̃=0.5)")

    out = joinpath(FIGURES_DIR, "q1-intermediate-depth.png")
    savefig(plt, out)
    println("wrote $out")
end

# ── E3: adiabatic-init optimization ────────────────────────────────────

function plot_adiabatic_init()
    path = joinpath(RESULTS_DIR, "q1-adiabatic-init.csv")
    isfile(path) || (println("skip E3 (no $path)"); return)
    header, rows = load_csv(path)

    Ds = sort(unique(asint(getcol(header, rows, "D"))))
    plt = plot(title="Adiabatic-initialised QAOA: best-of-grid vs warm-start",
               xlabel="D", ylabel="c̃",
               legend=:topright, xticks=Ds)

    best_vals  = Float64[]
    warm_vals  = Float64[]
    seed_bests = Float64[]
    for D in Ds
        sub = filter(r -> parse(Int, r[2]) == D, rows)
        cs  = asfloat(getcol(header, sub, "ctilde_seed"))
        cao = asfloat(getcol(header, sub, "ctilde_adi_opt"))
        cw  = asfloat(getcol(header, sub, "ctilde_warm"))[1]
        push!(best_vals,  maximum(cao))
        push!(warm_vals,  cw)
        push!(seed_bests, maximum(cs))
    end

    plot!(plt, Ds, warm_vals,  label="warm-start (c̃)",      lw=2, marker=:circle)
    plot!(plt, Ds, best_vals,  label="best adi-init (c̃)",   lw=2, marker=:utriangle)
    plot!(plt, Ds, seed_bests, label="best adi-seed (raw)", lw=1, ls=:dash, marker=:diamond)
    hline!(plt, [0.5], color=:gray, ls=:dot, label="random")

    out = joinpath(FIGURES_DIR, "q1-adiabatic-init.png")
    savefig(plt, out)
    println("wrote $out")
end

# ── E4: angle curvature (linear-fit r²) ────────────────────────────────

function plot_angle_curvature()
    path = joinpath(RESULTS_DIR, "q1-angle-curvature.csv")
    isfile(path) || (println("skip E4 (no $path)"); return)
    header, rows = load_csv(path)

    keep = filter(r -> parse(Int, r[5]) == 1, rows)  # deg = 1 only
    Ds   = sort(unique(asint(getcol(header, keep, "D"))))
    γ_r2 = Float64[]
    β_r2 = Float64[]
    for D in Ds
        sub_g = filter(r -> parse(Int, r[2]) == D && r[4] == "gamma", keep)
        sub_b = filter(r -> parse(Int, r[2]) == D && r[4] == "beta",  keep)
        push!(γ_r2, parse(Float64, sub_g[1][6]))
        push!(β_r2, parse(Float64, sub_b[1][6]))
    end

    plt = bar(Ds, γ_r2, label="γ", bar_position=:dodge,
              xlabel="D", ylabel="r²(linear fit)",
              title="Linear adiabatic fit quality vs D (deg=1)",
              ylim=(0, 1), xticks=Ds)
    bar!(plt, Ds .+ 0.35, β_r2, label="β", bar_width=0.35)
    hline!(plt, [1.0], color=:gray, ls=:dot, label="perfect linear")

    out = joinpath(FIGURES_DIR, "q1-angle-curvature.png")
    savefig(plt, out)
    println("wrote $out")
end

plot_angle_schedules()
plot_intermediate_depth()
plot_adiabatic_init()
plot_angle_curvature()

File: scripts/q1_plots.py
#!/usr/bin/env python3
"""Q1 plotting: reads the four Q1 CSVs and writes PNGs into figures/.

  q1-angle-schedules.csv      → figures/q1-angle-schedules.png   (E1)
  q1-intermediate-depth.csv   → figures/q1-intermediate-depth.png (E2)
  q1-adiabatic-init.csv       → figures/q1-adiabatic-init.png    (E3)
  q1-angle-curvature.csv      → figures/q1-angle-curvature.png   (E4)

Usage: python3 scripts/q1_plots.py
"""
from __future__ import annotations

import csv
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np

ROOT        = Path(__file__).resolve().parent.parent
RESULTS_DIR = ROOT / "results"
FIGURES_DIR = ROOT / "figures"
FIGURES_DIR.mkdir(exist_ok=True)


# ── CSV reader (skip lines starting with '#') ─────────────────────────

def load_csv(path: Path) -> tuple[list[str], list[dict]]:
    with path.open() as f:
        rows = [r for r in csv.reader(f) if r and not r[0].startswith("#")]
    header = rows[0]
    data   = [dict(zip(header, row)) for row in rows[1:]]
    return header, data


# ── E1: angle schedules ───────────────────────────────────────────────

def plot_angle_schedules() -> None:
    path = RESULTS_DIR / "q1-angle-schedules.csv"
    if not path.exists():
        print(f"skip E1 (no {path})")
        return
    _, rows = load_csv(path)
    Ds = sorted({int(r["D"]) for r in rows})

    fig, (axγ, axβ) = plt.subplots(1, 2, figsize=(14, 5), sharex=True)
    cmap = plt.get_cmap("viridis")

    for i, D in enumerate(Ds):
        sub = [r for r in rows if int(r["D"]) == D]
        sub.sort(key=lambda r: int(r["j"]))
        p   = int(sub[0]["p"])
        x   = np.array([int(r["j"]) for r in sub]) / p
        γo  = np.array([float(r["gamma_opt"]) for r in sub])
        βo  = np.array([float(r["beta_opt"])  for r in sub])
        γa  = np.array([float(r["gamma_adi"]) for r in sub])
        βa  = np.array([float(r["beta_adi"])  for r in sub])
        c   = cmap(i / max(1, len(Ds) - 1))
        axγ.plot(x, γo, "o-",  color=c, label=f"D={D} opt", lw=2)
        axγ.plot(x, γa, "--",  color=c, alpha=0.6, label=f"D={D} adi")
        axβ.plot(x, βo, "o-",  color=c, label=f"D={D} opt", lw=2)
        axβ.plot(x, βa, "--",  color=c, alpha=0.6, label=f"D={D} adi")

    for ax, title, ylabel in (
        (axγ, "γ schedule: optimal vs linear adiabatic", "γ"),
        (axβ, "β schedule: optimal vs linear adiabatic", "β"),
    ):
        ax.set_xlabel("step j / p")
        ax.set_ylabel(ylabel)
        ax.set_title(title)
        ax.grid(alpha=0.3)
        ax.legend(loc="best", fontsize=8, ncol=2)

    out = FIGURES_DIR / "q1-angle-schedules.png"
    fig.tight_layout()
    fig.savefig(out, dpi=150)
    plt.close(fig)
    print(f"wrote {out}")


# ── E2: intermediate-depth performance ────────────────────────────────

def plot_intermediate_depth() -> None:
    path = RESULTS_DIR / "q1-intermediate-depth.csv"
    if not path.exists():
        print(f"skip E2 (no {path})")
        return
    _, rows = load_csv(path)
    Ds = sorted({int(r["D"]) for r in rows})

    fig, ax = plt.subplots(figsize=(11, 6))
    cmap = plt.get_cmap("viridis")

    for i, D in enumerate(Ds):
        sub = sorted((r for r in rows if int(r["D"]) == D), key=lambda r: int(r["t"]))
        t   = np.array([int(r["t"])                       for r in sub])
        ct  = np.array([float(r["ctilde_truncated"])      for r in sub])
        co  = np.array([float(r["ctilde_optimal_at_t"])   for r in sub])
        ca  = np.array([float(r["ctilde_adiabatic_at_t"]) for r in sub])
        c   = cmap(i / max(1, len(Ds) - 1))
        ax.plot(t, co, "o-",  color=c, lw=2,  label=f"D={D} optimal@t")
        ax.plot(t, ct, "^--", color=c, lw=1,  alpha=0.8, label=f"D={D} truncated")
        ax.plot(t, ca, "d:",  color=c, lw=1,  alpha=0.7, label=f"D={D} adiabatic")

    ax.axhline(0.5, color="gray", ls=":", label="random (c̃=0.5)")
    ax.set_xlabel("depth t")
    ax.set_ylabel("c̃")
    ax.set_title("QAOA intermediate-depth performance")
    ax.set_ylim(0.45, 0.92)
    ax.grid(alpha=0.3)
    ax.legend(loc="lower right", fontsize=8, ncol=2)

    out = FIGURES_DIR / "q1-intermediate-depth.png"
    fig.tight_layout()
    fig.savefig(out, dpi=150)
    plt.close(fig)
    print(f"wrote {out}")


# ── E3: adiabatic-init optimization ───────────────────────────────────

def plot_adiabatic_init() -> None:
    path = RESULTS_DIR / "q1-adiabatic-init.csv"
    if not path.exists():
        print(f"skip E3 (no {path})")
        return
    _, rows = load_csv(path)
    if not rows:
        print(f"skip E3 (empty {path})")
        return
    Ds = sorted({int(r["D"]) for r in rows})

    best_adi  = []
    warm      = []
    seed_max  = []
    for D in Ds:
        sub = [r for r in rows if int(r["D"]) == D]
        best_adi.append(max(float(r["ctilde_adi_opt"]) for r in sub))
        warm.append(float(sub[0]["ctilde_warm"]))
        seed_max.append(max(float(r["ctilde_seed"]) for r in sub))

    fig, ax = plt.subplots(figsize=(9, 5.5))
    ax.plot(Ds, warm,     "o-", lw=2, label="warm-start (c̃)")
    ax.plot(Ds, best_adi, "^-", lw=2, label="best adi-init (c̃, after L-BFGS)")
    ax.plot(Ds, seed_max, "d--", lw=1, alpha=0.7, label="best adi-seed (c̃, raw)")
    ax.axhline(0.5, color="gray", ls=":", label="random")

    for D, w, b in zip(Ds, warm, best_adi):
        ax.annotate(f"Δ={w-b:+.3f}", (D, b), textcoords="offset points",
                    xytext=(0, -14), ha="center", fontsize=8, color="firebrick")

    ax.set_xticks(Ds)
    ax.set_xlabel("D")
    ax.set_ylabel("c̃")
    ax.set_title("Adiabatic-initialised QAOA: best-of-grid vs warm-start  (k=2, p=8)")
    ax.grid(alpha=0.3)
    ax.legend(loc="upper right")

    out = FIGURES_DIR / "q1-adiabatic-init.png"
    fig.tight_layout()
    fig.savefig(out, dpi=150)
    plt.close(fig)
    print(f"wrote {out}")


# ── E4: linear-fit r² and cubic improvement ───────────────────────────

def plot_angle_curvature() -> None:
    path = RESULTS_DIR / "q1-angle-curvature.csv"
    if not path.exists():
        print(f"skip E4 (no {path})")
        return
    _, rows = load_csv(path)

    deg1 = [r for r in rows if int(r["deg"]) == 1]
    deg3 = [r for r in rows if int(r["deg"]) == 3]
    Ds   = sorted({int(r["D"]) for r in deg1})

    γ_r2_lin, β_r2_lin = [], []
    γ_r2_cub, β_r2_cub = [], []
    for D in Ds:
        γ_r2_lin.append(float(next(r for r in deg1 if int(r["D"]) == D and r["profile"] == "gamma")["r2"]))
        β_r2_lin.append(float(next(r for r in deg1 if int(r["D"]) == D and r["profile"] == "beta") ["r2"]))
        γ_r2_cub.append(float(next(r for r in deg3 if int(r["D"]) == D and r["profile"] == "gamma")["r2"]))
        β_r2_cub.append(float(next(r for r in deg3 if int(r["D"]) == D and r["profile"] == "beta") ["r2"]))

    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(13, 5))
    x = np.arange(len(Ds))
    w = 0.35
    ax1.bar(x - w/2, γ_r2_lin, width=w, label="γ", color="C0")
    ax1.bar(x + w/2, β_r2_lin, width=w, label="β", color="C1")
    ax1.axhline(1.0, color="gray", ls=":", label="perfect linear")
    ax1.set_xticks(x); ax1.set_xticklabels(Ds)
    ax1.set_xlabel("D"); ax1.set_ylabel("r² (linear fit)")
    ax1.set_title("Linear adiabatic fit quality (deg=1)")
    ax1.set_ylim(0, 1.05); ax1.grid(alpha=0.3); ax1.legend()

    γ_dr2 = np.array(γ_r2_cub) - np.array(γ_r2_lin)
    β_dr2 = np.array(β_r2_cub) - np.array(β_r2_lin)
    ax2.bar(x - w/2, γ_dr2, width=w, label="γ", color="C0")
    ax2.bar(x + w/2, β_dr2, width=w, label="β", color="C1")
    ax2.set_xticks(x); ax2.set_xticklabels(Ds)
    ax2.set_xlabel("D"); ax2.set_ylabel("Δr² when going linear → cubic")
    ax2.set_title("Variance recovered by adding curvature")
    ax2.grid(alpha=0.3); ax2.legend()

    out = FIGURES_DIR / "q1-angle-curvature.png"
    fig.tight_layout()
    fig.savefig(out, dpi=150)
    plt.close(fig)
    print(f"wrote {out}")


def main() -> None:
    plot_angle_schedules()
    plot_intermediate_depth()
    plot_adiabatic_init()
    plot_angle_curvature()


if __name__ == "__main__":
    main()
